src/paillier.rs

```diff
@@ -78,3 +123,116 @@ impl PaillierDecryptionKey {
         PaillierEncryptionKey(libpaillier::EncryptionKey::from(&self.0))
     }
 }
+
+// Safe prime generation functions for production and testing.
+pub(crate) mod prime_gen {
+    use super::*;
+    #[cfg(test)]
+    use rand::Rng;
+    use rand::{CryptoRng, RngCore};
+
+    /// Sample a 512-bit safe prime uniformly at random.
+    ///
+    /// Prime size is derived from the security
+    /// parameter setting of κ = 128, and safe primes being of length 4κ (Figure 6,
+    /// Round 1 of the CGGMP'21 paper)
+    pub(crate) fn get_random_safe_prime_512<R: RngCore + CryptoRng>(rng: &mut R) -> BigNumber {
+        BigNumber::safe_prime_from_rng(PRIME_BITS, rng)
+    }
+
+    // Generate safe primes from a file. Usually, generating safe primes takes
+    // awhile (0-5 minutes per 512-bit safe prime on my laptop, average 50 seconds)
+    #[cfg(test)]
+    lazy_static::lazy_static! {
+        static ref POOL_OF_PRIMES: Vec<BigNumber> = get_safe_primes_from_file();
+    }
+
+    #[cfg(test)]
+    fn get_safe_primes_from_file() -> Vec<BigNumber> {
+        crate::safe_primes_512::SAFE_PRIMES
+            .iter()
+            .map(|s| BigNumber::from_slice(hex::decode(s).unwrap()))
+            .filter(|prime| prime.bit_length() == PRIME_BITS)
```

Comment:
note: I added this `filter`. Some of the primes in the list are not exactly 512 bits.

Reply:
Right, I saw the slack convo. Can you add a comment explaining the filter to the code?